src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -261,6 +261,93 @@ internal static ulong ToUInt64(object value)
         #endregion
 
         #region Public Static Methods
+        public static string? GetName<TEnum>(TEnum value) where TEnum : struct, Enum
+            => GetName(typeof(TEnum), value);
+
+        public static string? GetName(Type enumType, object value)
+        {
+            if (enumType is null)
+                throw new ArgumentNullException(nameof(enumType));
+
+            return enumType.GetEnumName(value);
+        }
+
+        public static string[] GetNames<TEnum>() where TEnum : struct, Enum
+            => GetNames(typeof(TEnum));
+
+        public static string[] GetNames(Type enumType)
+        {
+            if (enumType is null)
+                throw new ArgumentNullException(nameof(enumType));
+
+            return enumType.GetEnumNames();
+        }
+
+        internal static string[] InternalGetNames(RuntimeType enumType)
+        {
+            // Get all of the names
+            return GetEnumInfo(enumType, true).Names;
+        }
+
+        public static Type GetUnderlyingType(Type enumType)
+        {
+            if (enumType == null)
+                throw new ArgumentNullException(nameof(enumType));
+
+            return enumType.GetEnumUnderlyingType();
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
```

Comment:
The convention is to always keep `MethodImplOptions.InternalCall` definitions in the unshared partition, even if they happen to be same between CoreCLR and Mono. It makes rewriting more runtime code in C# a bit easier.